File: python/utm-parse/src/utm_parse/parser.py
"""
UTM URL Parser

Parse and normalize UTM parameters from URLs.
"""

from dataclasses import dataclass, field
from typing import Optional
from urllib.parse import urlparse, parse_qs, unquote_plus

# Platform click ID mappings
PLATFORM_CLICK_IDS = {
    "gclid": "google",
    "fbclid": "facebook",
    "msclkid": "bing",
    "ttclid": "tiktok",
    "li_fat_id": "linkedin",
}

# Non-standard source parameters (in priority order)
NON_STANDARD_SOURCE_PARAMS = [
    "ref",
    "referrer",
    "refer",
    "source",
    "via",
    "from",
    "origin",
    "src",
    "channel",
    "aff",
    "affiliate",
    "partner",
]


@dataclass
class UtmParams:
    """UTM parameters container."""
    utm_source: Optional[str] = None
    utm_medium: Optional[str] = None
    utm_campaign: Optional[str] = None
    utm_term: Optional[str] = None
    utm_content: Optional[str] = None
    referrer: Optional[str] = None
    platform: Optional[str] = None
    raw_param: Optional[str] = None

    def is_empty(self) -> bool:
        """Check if all UTM params are None."""
        return all(
            v is None
            for v in [
                self.utm_source,
                self.utm_medium,
                self.utm_campaign,
                self.utm_term,
                self.utm_content,
            ]
        )

    def to_dict(self) -> dict:
        """Convert to dictionary."""
        return {
            "utm_source": self.utm_source,
            "utm_medium": self.utm_medium,
            "utm_campaign": self.utm_campaign,
            "utm_term": self.utm_term,
            "utm_content": self.utm_content,
            "referrer": self.referrer,
            "platform": self.platform,
        }


def parse_utm_from_url(url: str) -> UtmParams:
    """
    Parse UTM parameters from a URL string.

    Args:
        url: URL string to parse

    Returns:
        UtmParams object with parsed values

    Example:
        >>> params = parse_utm_from_url("https://example.com?utm_source=google")
        >>> params.utm_source
        'google'
    """
    if not url:
        return UtmParams()

    params = UtmParams()

    try:
        parsed = urlparse(url)
        query_params = parse_qs(parsed.query)
    except Exception:
        return UtmParams()

    # Parse standard UTM parameters
    for key in ["utm_source", "utm_medium", "utm_campaign", "utm_term", "utm_content"]:
        values = query_params.get(key, [])
        if values:
            setattr(params, key, unquote_plus(values[0]))

    # If no utm_source, try platform click IDs
    if not params.utm_source:
        for param, source in PLATFORM_CLICK_IDS.items():
            if param in query_params:
                params.utm_source = source
                params.platform = source
                break

    # If still no utm_source, try non-standard params
    if not params.utm_source:
        for non_std in NON_STANDARD_SOURCE_PARAMS:
            values = query_params.get(non_std, [])
            if values:
                params.utm_source = unquote_plus(values[0])
                params.raw_param = non_std
                break

    return params


def normalize(params: UtmParams) -> UtmParams:
    """
    Normalize UTM params by trimming whitespace.

    Args:
        params: UtmParams object to normalize

    Returns:
        Normalized UtmParams object
    """
    def trim(val: Optional[str]) -> Optional[str]:
        if val is None:
            return None
        trimmed = val.strip()
        return trimmed if trimmed else None

    return UtmParams(
        utm_source=trim(params.utm_source),
        utm_medium=trim(params.utm_medium),
        utm_campaign=trim(params.utm_campaign),
        utm_term=trim(params.utm_term),
        utm_content=trim(params.utm_content),
        referrer=trim(params.referrer),
        platform=params.platform,
        raw_param=params.raw_param,
    )

File: rust/utm-parse/src/parser.rs
//! URL parsing for UTM parameters

use crate::error::UtmError;
use crate::params::{Source, UtmParams};
use crate::platform::source_from_click_id;

/// Maximum URL length to prevent DoS
const MAX_URL_LENGTH: usize = 8192;

/// Non-standard source parameters (in order of priority)
const NON_STANDARD_SOURCE_PARAMS: &[&str] = &[
    "ref",
    "referrer",
    "refer",
    "source",
    "via",
    "from",
    "origin",
    "src",
    "channel",
    "aff",
    "affiliate",
    "partner",
];

/// Parse UTM parameters from a URL string
pub fn parse_utm_from_url(url: &str) -> Result<UtmParams, UtmError> {
    // Validate URL
    if url.is_empty() {
        return Err(UtmError::EmptyUrl);
    }
    if url.len() > MAX_URL_LENGTH {
        return Err(UtmError::UrlTooLong);
    }

    // Find query string
    let query = match url.find('?') {
        Some(idx) => &url[idx + 1..],
        None => return Ok(UtmParams::new()),
    };

    let mut params = UtmParams::new();

    // Parse query string
    for pair in query.split('&') {
        let (key, value) = match pair.find('=') {
            Some(idx) => (&pair[..idx], &pair[idx + 1..]),
            None => (pair, ""),
        };

        // Decode percent-encoding
        let decoded_value = decode_percent(value);
        let decoded_key = decode_percent(key);

        match decoded_key.as_str() {
            // Standard UTM parameters
            "utm_source" => params.utm_source = Some(decoded_value),
            "utm_medium" => params.utm_medium = Some(decoded_value),
            "utm_campaign" => params.utm_campaign = Some(decoded_value),
            "utm_term" => params.utm_term = Some(decoded_value),
            "utm_content" => params.utm_content = Some(decoded_value),
            // Check for platform click IDs
            _ => {
                if let Some(source) = source_from_click_id(&decoded_key) {
                    if params.utm_source.is_none() {
                        params.utm_source = Some(source.as_str().to_string());
                        params.platform = Some(source);
                    }
                }
            }
        }
    }

    // If no standard UTM source, try non-standard parameters
    if params.utm_source.is_none() {
        for (key, value) in url.query_pairs() {
            let key_str = key.to_string();
            let value_str = value.to_string();

            // Check non-standard source params
            for &non_std in NON_STANDARD_SOURCE_PARAMS {
                if key_str == non_std && !value_str.is_empty() {
                    params.utm_source = Some(value_str.clone());
                    params.raw_param = Some(key_str);
                    return Ok(params);
                }
            }
        }
    }

    Ok(params)
}

/// Simple percent decoding
fn decode_percent(s: &str) -> String {
    let mut result = String::with_capacity(s.len());
    let mut chars = s.chars().peekable();

    while let Some(c) = chars.next() {
        if c == '%' {
            let hex: String = chars.by_ref().take(2).collect();
            if hex.len() == 2 {
                if let Ok(byte) = u8::from_str_radix(&hex, 16) {
                    result.push(byte as char);
                    continue;
                }
            }
            // Invalid encoding, keep the original
            result.push('%');
            result.push_str(&hex);
        } else if c == '+' {
            // + is commonly used for space in query strings
            result.push(' ');
        } else {
            result.push(c);
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_url() {
        assert!(parse_utm_from_url("").is_err());
    }

    #[test]
    fn test_no_query() {
        let params = parse_utm_from_url("https://example.com").unwrap();
        assert!(params.is_empty());
    }

    #[test]
    fn test_decode_percent() {
        assert_eq!(decode_percent("hello%20world"), "hello world");
        assert_eq!(decode_percent("%C3%A9"), "é");
    }
}
